packages/runtime/src/state/modules/element-trees.ts

```diff
@@ -0,0 +1,358 @@
+  return getElementTree(state, documentKey)?.elements ?? new Map()
+}
+
+export function getElementIds(state: State, documentKey: string): Map<string, string> {
+  return getElementTree(state, documentKey)?.elementIds ?? new Map()
+}
+
+export function getElement(state: State, documentKey: string, elementKey: string): Element | null {
+  return getElements(state, documentKey).get(elementKey) ?? null
+}
+
+export function getElementId(state: State, documentKey: string, elementKey: string): string | null {
+  return getElementIds(state, documentKey).get(elementKey) ?? null
+}
+
+export function reducer(state: State = getInitialState(), action: Action): State {
+  switch (action.type) {
+    case ActionTypes.CREATE_ELEMENT_TREE: {
+      const { document, descriptors } = action.payload
+      return new Map(state).set(document.key, buildElementTree(document.rootElement, descriptors))
+    }
+
+    case ActionTypes.DELETE_ELEMENT_TREE: {
+      const nextState = new Map(state)
+      const deleted = nextState.delete(action.payload.documentKey)
+      return deleted ? nextState : state
+    }
+
+    case ActionTypes.CHANGE_ELEMENT_TREE: {
+      const { oldDocument, newDocument, descriptors, operation } = action.payload
+      const documentKey = oldDocument.key
+      console.assert(
+        documentKey === newDocument.key,
+        `Mismatching document keys ${documentKey} !== ${newDocument.key}`,
+      )
+
+      const elementTree = state.get(documentKey)
+      if (elementTree == null) return state
+
+      const updatedElementTree = applyChanges(
+        elementTree,
+        descriptors,
+        { old: oldDocument.rootElement, new: newDocument.rootElement },
+        operation,
+      )
+
+      return new Map(state).set(documentKey, updatedElementTree)
+    }
+
+    default:
+      return state
+  }
+}
+
+export function* traverseElementTree(
+  element: Element,
+  descriptors: DescriptorsByComponentType,
+): Generator<Element> {
+  yield element
+  if (isElementReference(element)) return
+
+  const elementDescriptors = descriptors.get(element.type)
+  if (elementDescriptors == null) return
+
+  for (const [propKey, descriptor] of Object.entries(elementDescriptors)) {
+    const children = Introspection.getElementChildren(descriptor, element.props[propKey])
+    for (const child of children) {
+      yield* traverseElementTree(child, descriptors)
+    }
+  }
+}
+
+function getElementIdProp(
+  element: ElementData,
+  descriptors: DescriptorsByComponentType,
+): string | null {
+  const elementDescriptors = descriptors.get(element.type)
+  if (elementDescriptors == null) return null
+
+  for (const [propName, descriptor] of Object.entries(elementDescriptors)) {
+    const elementId = Introspection.getElementId(descriptor, element.props[propName])
+    if (elementId != null) return elementId
+  }
+
+  return null
+}
+
+export function buildElementTree(
+  rootElement: Element,
+  descriptors: DescriptorsByComponentType,
+): ElementTree {
```

Comment:
A rewrite of [`normalizeElement`](https://github.com/makeswift/makeswift/pull/894/files#diff-366e1d38abb10b78254fe441d335387f0bf21463b598e0cbfa26fbd4df140cd8L121).